# Exercise 4.23. Scalar QDA

## Intro

In this exercise, we will fit a QDA classifier in 1D data. QDA is all about breaking down the posterior p(y|x) into three components using Bayes's theorem. The three parameters are the class conditional p(x|y), the prior p(y) and the marginal likelihood p(x). 

QDA imposes that the class conditional be Gaussian. Also, the marginal likelihood 
can be decomposed into a summation of class conditionals and priors 
($p(x) = \sum p(y)(p(x|y)$). A good strategy to estimate the parameters of the Gaussian
and the prior is to use MLE, which is the main point of this exercise.

In [1]:
%load_ext autoreload
%autoreload 2
%matplotlib inline

from matplotlib import pyplot as plt
import numpy as np
from scipy.stats import norm

plt.rcParams['figure.figsize'] = 10, 8

## Load data

Males are encoded as 1 and females are encoded as 0.

In [2]:
x = np.asarray([67, 79, 71, 68, 67, 60], dtype=np.float32)
y = np.asarray([1, 1, 1, 0, 0, 0], dtype=np.bool)

## a

### Fit QDA model to data usign MLE

Remember from section 4.2.4 that the priors $\pi_c$ and the parameters 
($\mu_c$ and $\Sigma_c$) can be estimated separately.

In [3]:
pi_m = (y==1).sum()/len(y)
pi_f = (y==0).sum()/len(y)

mu_m = x[y==1].sum()/(y==1).sum()
mu_f = x[y==0].sum()/(y==0).sum()

Sigma_m = ((x[y==1] - mu_m) ** 2).sum() / (y==1).sum()
Sigma_f = ((x[y==0] - mu_f) ** 2).sum() / (y==0).sum()

sigma_m = Sigma_m ** 0.5
sigma_f = Sigma_f ** 0.5

print('pi_m = {0:.2f}\n'.format(pi_m))
print('pi_f = {0:.2f}\n'.format(pi_f))

print('mu_m = {0:.2f}\n'.format(mu_m))
print('mu_f = {0:.2f}\n'.format(mu_f))

print('Sigma_m = {0:.2f}\n'.format(sigma_m))
print('Sigma_f = {0:.2f}\n'.format(sigma_f))

pi_m = 0.50

pi_f = 0.50

mu_m = 72.33

mu_f = 65.00

Sigma_m = 4.99

Sigma_f = 3.56



### Remarks 

In the computation of the variance, we made use of the property that 
$(x - \mu)(x - \mu)^t = (x - \mu)^2$ when $x$ and $\mu$ are scalars.

## b

Plug-in prediction with the MLE parameters calculated in item a, for x = 72.

In [4]:
sample = 72
px_y_m = norm(loc=mu_m, scale=sigma_m).pdf(sample)
px_y_f = norm(loc=mu_f, scale=sigma_f).pdf(sample)
p_x = px_y_m * pi_m + px_y_f * pi_f

p_y_m_x = (px_y_m * pi_m) / p_x

print('p(y = m | x = 72, \\theta_MLE) = {0:.2f}'.format(p_y_m_x))

p(y = m | x = 72, \theta_MLE) = 0.83


## c

If this was a multidimensional problem, we could not make use of $(x - \mu)(x - \mu)^t = (x - \mu)^2$. This is the only difference from the 1D case for the multidimensional case.

## Conclusion

In this exercise we perform scalar QDA. We used MLE to fit the parameters. In 
item b, $p(y = m | x = 72, \theta_{MLE}) = 0.83$, which makes a lot of sense
because x = 72 is very close to the mean of c = male and far away from the mean of c = female. This approach could be easily adapted to the multidimensional case. We only need to be careful in the computation of the covariance matrix.